### DISCUSSION TITLE
No Java classes available in tycho project within eclipse

### DISCUSSION THREAD

**Discussion by moley** [eclipse-tycho/tycho#1629] [Q&A] [answered]
Hi to all, 

I have a tycho build, which is working in commandline (means mvn verify, mvn install and mvn eclipse:eclipse works like a charme). 
As soon as I import the project via "Import - Existing mavenproject" (after .project and .classpath files are removed) into eclipse (2022-09, Build id: 20220908-1902) on mac it seems that eclipse does not detect any sourcefolder. No classes can be resolved. As well no JDT builder is configured. As soon as I configure Java Nature manually JDT builder is triggered, but I recieve many build errors (e.g. package seems to be not correct although in Java Build Path properties folder "src" is detected correctly as source folder. Can you please help me? What's the default way of opening a tycho build inside eclipse?

**Comment by jcompagner** [comment]
Can I ask, why do you remove .project and .classpath. The 2 files that have this info/configuration for eclipse?

**Comment by moley** [comment]
Coming from idea and gradle I thought this is the way to go. Because in this scenario buildship plugin recreates the files. So is the right way here to create .project and .classpath with mvn and import as mvn project as second step?

**Comment by mickaelistria** [comment, marked as answer]
Tycho projects are usually Eclipse PDE projects first, and most of PDE metadata is stored in .project and .classpath. Tycho is MANIFEST-first, you'd rather actually keep the .project and .classpath and use Tycho pom-less which is capable of dynamically generating the pom files (so you don't have to store them in your Git repo).
The Eclipse-first approach works more easily than the Maven-first approach for PDE projects, both in the IDE and in Tycho.

**Reply by laeubi** [reply]
Project files itself are not that essential but `.classpath` often is because the JDT(!) default is to use different layout than maven and most of the time this is not reflected in the pom configuration.... I think I already created a bug for this but is lost somehwere at all the platform repositories (can't find it anymore)

**Comment by moley** [comment]
Ok, thanks for all the feedback. I will try it out. Maybe a user documentation at a prominent place would be nice? Or - if yet available - a link in the github README.md. If available and I didn't see, then sorry about me

**Comment by HannesWell** [comment]
What has been said is all correct, but I want to add that recently the previously separate Tycho-Connector was migrated to M2E
- https://github.com/eclipse-m2e/m2e-core/pull/967

TLDR; unless your project requires special `.classpath` files, M2E (the Maven integration of Eclipse) should generate all the mentioned files for Tycho Maven projects as well with the upcoming release (will probably be part of Eclipse 2022-12).

If you are interested you can try out the latest M2E 2.1.0 snapshot at:
https://download.eclipse.org/technology/m2e/snapshots/2.1.0/

If there is something missing and you think it is a general issue, please open a dedicated issue at [Eclipse M2E](https://github.com/eclipse-m2e/m2e-core/) and consider to contribute a corresponding patch. :)

**Comment by lppedd** [comment]
Imho one of the problems of Eclipse is that it hides Maven. You tend to forget about it with all the autoconfiguration. Also some parts of Tycho extract info from Eclipse-first configuration files, but that isn't obvious.

I prefer to work directly with Maven and that's why it has been a mission for me to get IDEA working with Tycho customizations.

**Reply by laeubi** [reply]
This does not makes much sense, as Tycho is all about reusing the provided meta-data generated from the IDE :-)

If you like to work directly with Maven you don't need Tycho at all, just dont forget to configure the jar plugin to include your manifest or generate it automatically. Tycho itself do it that way already.

**Reply by lppedd** [reply]
@laeubi unfortunately I have seen first hand what happens with Eclipse + Tycho. Once Tycho is in place it's completely forgotten about. I mean, it's fine, as you get a transparent PDE development experience, but the Maven part of the codebase tends to be neglected over time, and that doesn't happen with a standard Maven-first environment.

> you don't need Tycho at all

That's a big statement 😄, without Tycho I would go crazy with configuring the build system. I mean I'm lucky it exists no joke.
But the fact it overrides lifecycles, provides new packaging types, extensions and so on require the IDE to know about it (see M2E for Tycho for example).

I'm not criticizing Tycho, I'm just exposing what I've seen happen around and certainly not in small companies.

**Reply by lppedd** [reply]
> Tycho is all about reusing the provided meta-data generated from the IDE

Forgot to add my 2c about this. This means you're tying a build system to an IDE. It's fine if you like the IDE it supports. It's not fun if you like alternatives.

**Reply by laeubi** [reply]
>  But the fact it overrides lifecycles, provides new packaging types, extensions and so on require the IDE to know about it (see M2E for Tycho for example).

That's what I want to express your are looking at it from the wrong side of the coin. It is not that Tycho defines anything and the IDE must use it, it is the IDE that defines things and Tycho allows you to use that same stuff inside a maven build. And if you use pomless you literally need no "maven" for this beside the master pom that configures some of your mojos if you like.

**Reply by lppedd** [reply]
@laeubi 

> And if you use pomless

I mean, pomless is one of the thing I would never use. It generates Maven metadata at build time anyway AFAIR.
But what I mean is **all** IDEs support Maven as a build system. Maven is used to compute dependencies that are later injected into the IDE workspace model.

Being that Tycho changes how Maven works, supporting it in IDEs is a pain.

> Tycho allows you to use that same stuff inside a maven build

But wouldn't a build system prefer to be IDE agnostic? Maven is definitely IDE agnostic per-se, that's why it's mostly easy to support usually. If an extension ties Maven to a tool, then Maven loses one of its strong point and every other tool must align their codebases.

**Reply by laeubi** [reply]
> I mean, pomless is one of the thing I would never use. It generates Maven metadata at build time anyway AFAIR.

No one forces you, but that perfectly shows how Tycho works, it does not need maven, it needs the input from the IDE only to work. So if one don't want this one should not use maven, one should always use the right tool for the right task.

> But what I mean is all IDEs support Maven as a build system.
> Being that Tycho changes how Maven works, supporting it in IDEs is a pain.

Tycho is a maven plugin, so if it does not work in your "maven supporting ide" then this IDE does not support maven, Tycho is not doing anything special, forbidden or different than what is "maven". But I think we have discussed this before already.

> But wouldn't a build system prefer to be IDE agnostic? 

Tycho was made to bring the IDE tools to the commandline so you can use your ide (that does not know anything about maven) and build it on the commandline with maven.

> Maven is definitely IDE agnostic per-se that's why it's mostly easy to support usually

Maven is a commandline build tool, it has very limited support for IDEs and their devolopers even refuse to improve that. And the tycho (and many others plugins as well) proof that it is actually not "easy" to support maven. It is easy for the easy cases but fully supporting maven in way users can benefit from it in an IDE is not.

**Reply by lppedd** [reply]
@laeubi 

> But I think we have discussed this before already

Yup we have. Definitely I could be biased because I'm used to work on environments that are easily cross-IDE and I can pick what I think it's comfortable for me.  
The "closed"/niche nature of PDE/p2 is what makes it difficult to use. My main goal in expressing my opinion is to improve the overall development experience. Because as of now developing Eclipse-based products is really possible only on Eclipse, all the other IDEs require some sort of hacks. This hits to a point, no joke, I tend to avoid Eclipse related jobs if I'm not sure I can workaround on other tools, and definitely I'm not the only one.  
The fact Tycho is built over Maven helps a ton, because Maven is mostly supported everywhere and I can thinker around open source IDEs and build my own if required (that's what I'm doing with IJ). Without Tycho, Eclipse would be really the only tool to use.

**Reply by laeubi** [reply]
>  Because as of now developing Eclipse-based products is really possible only on Eclipse

Nowdays it is possible to develop Linux tools on Windows (and the other way round), still I would use Linux in case I need to develop Linux Apps and Windows when developing Windows apps. So I always choose the best tool for the job.

And actually it is InteliJ not supporting PDE that makes it really hard to develop Eclipse Based software without Eclipse ... compiling/packaging is just a very very limited part of that (and that what Tycho is actually good at). So when I develop Eclipse Applications I use Eclipse and PDE, i *never* run maven/tycho ... it just works on the CI (at least that's the goal of Tycho).

**Reply by lppedd** [reply]
@laeubi 

> So I always choose the best tool for the job

But best tool is subjective, so being open to a wider audience is a way to have more people around.
I don't hate Eclipse, but I don't like it either and would pick something else to spend my day on. I like RCP tho, it's the best framework for desktop applications around.

> And actually it is IntelliJ not supporting PDE

It's been like that for a long time, and it's valid for other tools too. My first encounter with PDE was back in 2015 and the progress I've seen in tooling, apart from Eclipse itself obviously, has been minimal. When I've worked with other platforms, especially web and mobile, I've always seen a lot of communication between tools makers and frameworks makers. But the input has always come from the framework first.

> So when I develop Eclipse Applications I use Eclipse and PDE, i never run maven/tycho

Me too with IJ. You can see that thanks to Tycho injecting dependencies into the Maven model I have fully functioning environment. I spend two minutes importing the project as a Maven project, and then I code like any other Java project. Sure, I can't use WindowBuilder for example, or I don't have a way to attach sources automatically yet, but the fact I can do my favorite devops stuff + standalone coding (in Kotlin too!) in a single place is a lot of gained productivity. I don't want to open Eclipse exclusively to develop plugins.

![image](https://user-images.githubusercontent.com/19871649/200182319-24566732-f363-482f-a8f0-dbb80fa1bb7b.png)

**Reply by laeubi** [reply]
> But the input has always come from the framework first.

Well if the framework offers an IDE why should it give input to other IDEs how to support this framework better if there is no interest in it? Eclipse is offered for free and many people do work on it for free, so there always not much time left to contribute to a completely different eco system. So I think this is mostly a combination of less contribution and less need for InteliJ that it do not makes much progress. Also it does not seem someone wants to spend money on improve the PDE development... Even though it would not be that hard to support IntelliJ.

Where it comes to "maven > eclipse (with tycho)" m2e is more suitable place for improvements, but again ... there is too much work for too less developers... so you actually have only limited options:

1. Contribute your self
2. Hope someone else is bugged enough and does
3. pay some one (for example me) to do it if 1 or 2 do not suffice

**Reply by lppedd** [reply]
@laeubi 

> if there is no interest in it

Every OS initiative starts on the 90-9-1 model, where only the 1% do something. When the engagement start going up, contributors start arriving too. The fact PDE is not being addressed in other tools is probably more of a political thing from every perspective:
1. Eclipse does not try to push for it
2. Not many users in other IDEs as it's frustrating/impossible
3. IDE makers have no interest because the userbase is too small

This is a infinite loop unfortunately.

> there is too much work for too less developers

That's perfectly understandable and I'm not here to ask for full support in IJ. On the Tycho side it's just an added cool thing to remember about other IDEs during development. E.g. if someday you decide to support only POM-less, IJ users are definitely screwed.

> Contribute your self

Yeah that's what I did, there is an open PR for Tycho support in IJ on the community repository

**Reply by jcompagner** [reply]
isn't tycho really a build tool to build eclipse based or also maybe "osgi" based applications?

You seem to want to manage dependencies with mvn for osgi based stuff?

I would really hate that i need to also specify my eclipse projects/plugins also in mvn, thats not where i say where it depends on
thats all in the manifest.. (package imports, or bundle imports)

The system (and thats what tycho really does when building command line) is setting up all the dependencies based on target platform (and i think that target platform can be configured more and more to come from pure mvn central artifacts, as long as they are correct osgi bundles)

But you seem to be thinking this as pure java projects with dependencies that are coming from somewhere like a pom.xml file..

I am not sure what you are building, but if you are building a Eclipse based application then there you just have to use tycho to build it fully so that you fuse mvn and osgi together. 
But if you are just building an (web)application and you are not really building a eclipse based application you shouldn't use tycho at all.. then you can just just use "pure" mvn.

**Reply by HannesWell** [reply]
If you want to build OSGi applications in a pom/maven first approach you also may want to consider to use the `bnd-maven-plugin` to only generate you manifest.
- https://mvnrepository.com/artifact/biz.aQute.bnd/bnd-maven-plugin 
- https://github.com/bndtools/bnd/tree/master/maven/bnd-maven-plugin

Your desire to have a pom/maven first and only build is totally valid and I agree that this is a way that is more cross-IDE compatible.
Nevertheless to goal of Tycho is actually to fully support and leverage the `META-INF/MANIFEST.MF`-first approach used/introduced by Eclipse PDE. And at least my goal when it comes to Tycho is actually always to reduce the number of pom-files as far as possible and let Tycho derive as much as possible from the other PDE files in order to avoid duplication of information.

If you have suggestions for certain improvements please don't hesitate to mention them, but steering Tycho towards using more poms will likely not happen because it contradicts the goal of Tycho to be a 'Manifest-first' build tool.
Nevertheless you are still not forced to use Tycho without poms. And in Tycho 2.7.5 there even has been a `generate-poms` mojo:
https://tycho.eclipseprojects.io/doc/2.7.5/tycho-pomgenerator-plugin/generate-poms-mojo.html

**Reply by lppedd** [reply]
@HannesWell thanks for the suggestion. I had starred that plugin some time ago while I was evaluating possible approaches in keeping the IDE-related metadata at a minimum. To be honest in the OSGi ecosystem there is currently nothing comparable to what Tycho offers, and that's why I'm still very comfortable with using it. Otherwise I would not have opened a PR or I would not be following issues/PRs daily.

> reduce the number of pom-files as far as possible

This is a something I perfectly understand when looking at it from an Eclipse-first development experience. However I do hope the time of pom-less-only never arrives, or at least not before the tooling in the PDE ecosystem has improved.

It's not really that I *want* a pure Maven solution, it's more that Tycho and PDE are developed in parallel because of obvious reasons, and other development environments do not get the necessary attention. In the past six months I was able to reduce the interactions with Eclipse as far as not having to open it to launch the application. I got Tycho to load correctly in IntelliJ IDEA, but the hours I've spent on these points are just surreal.

> steering Tycho towards using more poms

No no, to clarify, I don't want it to use *more* POMs, just the POMs required by the Maven standard. If Tycho contributes to the Maven model with properties, dependencies, and whatsoever, keeping it as standard as possible then refactoring IDE Maven integrations to fully cover what Tycho does is possible.

Some people here see Maven as only a build tool, but it is extremely important for other IDEs as they can actually allow coding without understanding what OSGi/PDE is, they just need to know about Maven.

**Reply by laeubi** [reply]
> No no, to clarify, I don't want it to use _more_ POMs, just the POMs required by the Maven standard. If Tycho contributes to the Maven model with properties, dependencies, and whatsoever, keeping it as standard as possible then refactoring IDE Maven integrations to fully cover what Tycho does is possible.
> 
> Some people here see Maven as only a build tool, but it is extremely important for other IDEs as they can actually allow coding without understanding what OSGi/PDE is, they just need to know about Maven.

I think there is a general misunderstanding here, pomless build do not mean there is no pom nor that it is not "maven standard", it is based on [polyglot-extension](https://github.com/takari/polyglot-maven) and every "maven aware ide" should be able to support this.

**Comment by laeubi** [comment]
@moley I think we have diverged a bit from the original topic. For your particular case, if you think things can be improved it would be best to share an example that reproduces the problem and head over to [m2eclipse](https://github.com/eclipse-m2e/m2e-core) what actually do the maven > eclipse conversion here.

**Comment by moley** [comment]
Yes, sorry for starting this discussion. Didn't want to blame or grumble. 
So meanwhile I am step ahead.... Projects are imported and src folders are used. I have another problem recieving "Project is missing required Java Project" although compilable at commandline, I can head over to m2eclipse for this case. Thanks for leading me to the right direction 👍
